File: tools/libhm64/sprites/addresses.py
"""
Sprite address parsing and lookup utilities.

Parses the sprite_addresses.csv file and provides cached lookup of sprite info.
"""

import csv
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional

from ..common import rom
from ..data import SPRITE_ADDRESSES_CSV

_REPO_DIR = Path(__file__).resolve().parent.parent.parent.parent
DEFAULT_CSV_PATH = SPRITE_ADDRESSES_CSV
DEFAULT_OUTPUT_DIR = _REPO_DIR / "assets" / "sprites"

# Cached data
_sprites: Optional[List['SpriteInfo']] = None
_label_index: Optional[Dict[str, 'SpriteInfo']] = None


@dataclass
class SpriteInfo:
    """Information about a sprite asset from the CSV."""
    sprite_type: str  # 'type-1' or 'type-2'
    addr_base: int
    addr_index: int
    label: str
    subdir: str
    # Type-1 specific
    spritesheet_index_base: Optional[int] = None
    spritesheet_index_end: Optional[int] = None
    # Type-2 specific
    addr_extra: Optional[int] = None
    # Row index in CSV (for animation utilities compatibility)
    csv_row: int = 0

    @property
    def is_type1(self) -> bool:
        return self.sprite_type == 'type-1'

    @property
    def is_type2(self) -> bool:
        return self.sprite_type == 'type-2'


def _load_sprites(csv_path: Path = DEFAULT_CSV_PATH) -> None:
    """Load and cache sprite data from CSV."""
    global _sprites, _label_index

    _sprites = []
    _label_index = {}

    with open(csv_path, newline='', encoding='utf-8') as f:
        reader = csv.reader(f)
        for row_idx, row in enumerate(reader):
            row = [c.strip() for c in row]

            if len(row) == 6:
                # Type-1: base, index, spritesheet_index_base, spritesheet_index_end, label, subdir
                info = SpriteInfo(
                    sprite_type='type-1',
                    addr_base=int(row[0], 16),
                    addr_index=int(row[1], 16),
                    spritesheet_index_base=int(row[2], 16),
                    spritesheet_index_end=int(row[3], 16),
                    label=row[4],
                    subdir=row[5],
                    csv_row=row_idx
                )
            elif len(row) == 5:
                # Type-2: base, index, extra, label, subdir
                info = SpriteInfo(
                    sprite_type='type-2',
                    addr_base=int(row[0], 16),
                    addr_index=int(row[1], 16),
                    addr_extra=int(row[2], 16),
                    label=row[3],
                    subdir=row[4],
                    csv_row=row_idx
                )
            else:
                continue

            _sprites.append(info)
            _label_index[info.label] = info


def get_all_sprites(csv_path: Path = DEFAULT_CSV_PATH) -> List[SpriteInfo]:
    """Get list of all sprite info from CSV."""
    if _sprites is None:
        _load_sprites(csv_path)
    return _sprites


def get_sprite_by_label(label: str, csv_path: Path = DEFAULT_CSV_PATH) -> Optional[SpriteInfo]:
    """Get sprite info by label name."""
    if _label_index is None:
        _load_sprites(csv_path)
    return _label_index.get(label)


def clear_cache() -> None:
    """Clear cached sprite data."""
    global _sprites, _label_index
    _sprites = None
    _label_index = None


def get_asset_offsets(addr_index: int) -> List[int]:
    """Read 8 u32 offsets from asset index."""
    return [rom.read_u32_be(addr_index + i * 4) for i in range(8)]


def get_spritesheet_offsets(info: SpriteInfo) -> List[int]:
    """Get spritesheet offset table for a sprite."""
    if info.is_type1:
        return _get_spritesheet_offsets_type1(info)
    else:
        return _get_spritesheet_offsets_type2(info)


def _get_spritesheet_offsets_type1(info: SpriteInfo) -> List[int]:
    """Get spritesheet offset table for type-1 sprites."""
    base = info.spritesheet_index_base
    end = info.spritesheet_index_end
    count = (end - base) // 4
    offsets = [rom.read_u32_be(base + i * 4) for i in range(count)]

    # Special case: festivalFlowers has embedded spritesheet index at start of texture
    if info.label == 'festivalFlowers':
        embedded_index_size = count * 4
        last_nonzero = len(offsets) - 1
        while last_nonzero >= 0 and offsets[last_nonzero] == 0:
            last_nonzero -= 1
        offsets = [
            off + embedded_index_size if i <= last_nonzero else 0
            for i, off in enumerate(offsets)
        ]

    return offsets


def _get_spritesheet_offsets_type2(info: SpriteInfo) -> List[int]:
    """Get spritesheet offset table for type-2 sprites."""
    base = info.addr_base
    first_offset = rom.read_u32_be(base)

    # Validate: first_offset should be reasonable (< 64KB) and aligned
    if first_offset == 0 or first_offset > 0x10000 or first_offset % 4 != 0:
        return []

    count = first_offset // 4
    # Sanity check: max 1024 sprites per asset
    if count > 1024:
        return []

    return [rom.read_u32_be(base + i * 4) for i in range(count)]


def get_palette_offsets(addr_base: int, asset_offsets: List[int]) -> List[int]:
    """Get palette offset table."""
    palette_section = addr_base + asset_offsets[1]
    first_offset = rom.read_u32_be(palette_section)

    # Validate: first_offset should be reasonable (< 4KB) and aligned
    if first_offset == 0 or first_offset > 0x1000 or first_offset % 4 != 0:
        return []

    count = first_offset // 4
    # Sanity check: max 256 palettes
    if count > 256:
        return []

    return [rom.read_u32_be(palette_section + i * 4) for i in range(count)]


def get_animation_offsets(addr_base: int, asset_offsets: List[int]) -> List[int]:
    """Get animation offset table."""
    anim_section = addr_base + asset_offsets[2]
    first_offset = rom.read_u32_be(anim_section)

    if first_offset == 0:
        return []

    count = first_offset // 4
    return [rom.read_u32_be(anim_section + i * 4) for i in range(count)]


def get_sprite_to_palette_map(addr_base: int, asset_offsets: List[int], sprite_count: int) -> List[int]:
    """Get sprite-to-palette mapping array."""
    map_base = addr_base + asset_offsets[3] + 4  # Skip header
    rom_data = rom.get_rom()
    return list(rom_data[map_base:map_base + sprite_count])

File: tools/modding/sprite_editor/src/extract/addresses.rs
//! Embedded parser for `tools/libhm64/data/sprite_addresses.csv`. Frozen
//! original-ROM data — see memory note "frozen asset catalogues".
//!
//! The CSV has two row shapes — see `libhm64/sprites/addresses.py`:
//! - 7 columns (type-1): addr_base, addr_index, spritesheet_index_base,
//!   spritesheet_index_end, label, subdir, kind
//! - 6 columns (type-2): addr_base, addr_index, addr_extra, label, subdir, kind

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SpriteAddress {
    pub addr_base: u32,
    pub addr_index: u32,
    pub label: String,
    pub subdir: String,
    pub kind: String,
    pub variant: AddressVariant,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AddressVariant {
    /// Type-1: spritesheet index is a separate region in ROM.
    Type1 { spritesheet_index_base: u32, spritesheet_index_end: u32 },
    /// Type-2: spritesheet index is embedded; `addr_extra` points to the next region.
    Type2 { addr_extra: u32 },
}

#[derive(Debug, thiserror::Error)]
pub enum AddressesError {
    #[error("line {line}: invalid hex `{value}` in column `{column}`")]
    BadHex { line: usize, value: String, column: &'static str },
}

pub fn parse(text: &str) -> Result<Vec<SpriteAddress>, AddressesError> {
    let mut out = Vec::new();
    for (i, line) in text.lines().enumerate() {
        if line.trim().is_empty() {
            continue;
        }
        let cols: Vec<String> = line.split(',').map(|s| s.trim().to_string()).collect();
        let line_no = i + 1;
        let parse_hex = |val: &str, col: &'static str| {
            u32::from_str_radix(val, 16).map_err(|_| AddressesError::BadHex {
                line: line_no,
                value: val.to_string(),
                column: col,
            })
        };
        let row = match cols.len() {
            7 => SpriteAddress {
                addr_base: parse_hex(&cols[0], "addr_base")?,
                addr_index: parse_hex(&cols[1], "addr_index")?,
                label: cols[4].clone(),
                subdir: cols[5].clone(),
                kind: cols[6].clone(),
                variant: AddressVariant::Type1 {
                    spritesheet_index_base: parse_hex(&cols[2], "spritesheet_index_base")?,
                    spritesheet_index_end: parse_hex(&cols[3], "spritesheet_index_end")?,
                },
            },
            6 => SpriteAddress {
                addr_base: parse_hex(&cols[0], "addr_base")?,
                addr_index: parse_hex(&cols[1], "addr_index")?,
                label: cols[3].clone(),
                subdir: cols[4].clone(),
                kind: cols[5].clone(),
                variant: AddressVariant::Type2 { addr_extra: parse_hex(&cols[2], "addr_extra")? },
            },
            // Match the Python loader: silently skip rows that aren't 6 or 7 columns.
            _ => continue,
        };
        out.push(row);
    }
    Ok(out)
}

pub fn embedded() -> Result<Vec<SpriteAddress>, AddressesError> {
    parse(include_str!("../../../../libhm64/data/sprite_addresses.csv"))
}

/// Read the 8 u32-BE asset offsets at `addr_index`. Mirrors
/// `addresses.py::get_asset_offsets`.
pub fn read_asset_offsets(rom: &[u8], addr_index: u32) -> [u32; 8] {
    let mut out = [0u32; 8];
    for (i, slot) in out.iter_mut().enumerate() {
        let off = addr_index as usize + i * 4;
        *slot = crate::io::binary::read_u32_be(rom, off).unwrap_or(0);
    }
    out
}

/// Build the spritesheet offset table for a sprite. Mirrors
/// `addresses.py::get_spritesheet_offsets`.
pub fn spritesheet_offsets(rom: &[u8], info: &SpriteAddress) -> Vec<u32> {
    match &info.variant {
        AddressVariant::Type1 { spritesheet_index_base, spritesheet_index_end } => {
            let count = (spritesheet_index_end - spritesheet_index_base) as usize / 4;
            let mut offsets: Vec<u32> = (0..count)
                .map(|i| {
                    let off = *spritesheet_index_base as usize + i * 4;
                    crate::io::binary::read_u32_be(rom, off).unwrap_or(0)
                })
                .collect();
            if info.label == "festivalFlowers" {
                let embedded_index_size = (count * 4) as u32;
                let last_nonzero = offsets.iter().rposition(|&v| v != 0).map(|p| p as i32).unwrap_or(-1);
                for (i, off) in offsets.iter_mut().enumerate() {
                    if (i as i32) <= last_nonzero {
                        *off += embedded_index_size;
                    } else {
                        *off = 0;
                    }
                }
            }
            offsets
        }
        AddressVariant::Type2 { .. } => {
            let base = info.addr_base as usize;
            let first_offset = crate::io::binary::read_u32_be(rom, base).unwrap_or(0);
            if first_offset == 0 || first_offset > 0x10000 || first_offset % 4 != 0 {
                return Vec::new();
            }
            let count = first_offset as usize / 4;
            if count > 1024 {
                return Vec::new();
            }
            (0..count)
                .map(|i| crate::io::binary::read_u32_be(rom, base + i * 4).unwrap_or(0))
                .collect()
        }
    }
}

/// Build the palette offset table.
pub fn palette_offsets(rom: &[u8], addr_base: u32, asset_offsets: &[u32; 8]) -> Vec<u32> {
    if asset_offsets[1] == 0 {
        return Vec::new();
    }
    let palette_section = (addr_base + asset_offsets[1]) as usize;
    let first_offset = crate::io::binary::read_u32_be(rom, palette_section).unwrap_or(0);
    if first_offset == 0 || first_offset > 0x1000 || first_offset % 4 != 0 {
        return Vec::new();
    }
    let count = first_offset as usize / 4;
    if count > 256 {
        return Vec::new();
    }
    (0..count)
        .map(|i| crate::io::binary::read_u32_be(rom, palette_section + i * 4).unwrap_or(0))
        .collect()
}

/// Build the animation offset table.
pub fn animation_offsets(rom: &[u8], addr_base: u32, asset_offsets: &[u32; 8]) -> Vec<u32> {
    let anim_section = (addr_base + asset_offsets[2]) as usize;
    let first_offset = crate::io::binary::read_u32_be(rom, anim_section).unwrap_or(0);
    if first_offset == 0 {
        return Vec::new();
    }
    let count = first_offset as usize / 4;
    (0..count)
        .map(|i| crate::io::binary::read_u32_be(rom, anim_section + i * 4).unwrap_or(0))
        .collect()
}

/// Read the sprite-to-palette mapping bytes (one u8 per sprite).
pub fn sprite_to_palette_map(
    rom: &[u8],
    addr_base: u32,
    asset_offsets: &[u32; 8],
    sprite_count: usize,
) -> Vec<u8> {
    let map_base = (addr_base + asset_offsets[3]) as usize + 4; // skip 4-byte header
    rom.get(map_base..map_base + sprite_count).map(|s| s.to_vec()).unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_type1_row() {
        let csv = "159330,2D0660,2D0680,2D15E0,player,entitySprites,entity\n";
        let rows = parse(csv).unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0].addr_base, 0x159330);
        assert_eq!(rows[0].addr_index, 0x2D0660);
        assert_eq!(rows[0].label, "player");
        assert_eq!(rows[0].kind, "entity");
        assert!(matches!(rows[0].variant, AddressVariant::Type1 { .. }));
    }

    #[test]
    fn parses_type2_row() {
        let csv = "7FCD10,7FD1E0,7FD200,shadows,entitySprites,standalone\n";
        let rows = parse(csv).unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0].label, "shadows");
        assert!(matches!(rows[0].variant, AddressVariant::Type2 { .. }));
    }

    #[test]
    fn rejects_bad_hex() {
        let csv = "ZZZZ,2D0660,2D0680,2D15E0,player,entitySprites,entity\n";
        let err = parse(csv).unwrap_err();
        assert!(matches!(err, AddressesError::BadHex { .. }));
    }

    #[test]
    fn embedded_csv_parses() {
        let rows = embedded().expect("embedded CSV must parse");
        assert!(!rows.is_empty());
        assert!(rows.iter().any(|r| r.label == "player"));
        assert!(rows.iter().any(|r| r.label == "shadows"));
    }
}
